### PR TITLE
exp: output troubleshooting link on shallow merge failure

### LINKED ISSUE #7547: dvc exp run unavailable in CML
Is there any way to run DVC exp run in CML?

My use case includes training several experiments in a Cloud or self hosted Server and publish in the PR the results of my work. 

For now when trying to do this using the setup action I get an unexpected error.

I understand that commit hides experiment results and I read that CML is not intended for experimentation but that is not quite clear in official documentation.

### PR DESCRIPTION
**Pull request by pmrowla**
* [x] ❗ I have followed the [Contributing to DVC](https://dvc.org/doc/user-guide/contributing/core) checklist.

* [x] 📖 If this PR requires [documentation](https://dvc.org/doc) updates, I have created a separate PR (or issue, at least) in [dvc.org](https://github.com/iterative/dvc.org) and linked it here.

Thank you for the contribution - we'll try to review it as soon as possible. 🙏

Will close #7547 

Depends on https://github.com/iterative/scmrepo/pull/75

```
$ dvc exp run
ERROR: Merge analysis failed: Cannot merge in shallow Git repo. See <https://dvc.org/doc/user-guide/troubleshooting#git-shallow> for more information.
```

### CODE REVIEW

File: dvc/scm.py [hunk 1]
```diff
@@ -47,6 +49,25 @@ def __init__(self, reason: str) -> None:
...
 
+class GitMergeError(SCMError):
+    def __init__(self, msg: str, scm: Optional["Git"] = None) -> None:
+        if scm and self._is_shallow(scm):
+            url = format_link(
+                "https://dvc.org/doc/user-guide/troubleshooting#git-shallow"
+            )
+            msg = (
+                f"{msg}: Cannot merge in shallow Git repo. "
+                f"See {url} for more information."
```
Comment (jorgeorpinel) [inline dvc/scm.py:60]: Just to clarify, is this a problem beyond `dvc exp` commands? Why are there internal Git merges again? Just FMI + maybe there's a better way to explain the issue (once I fully understand it). Thanks

Comment (pmrowla) [reply dvc/scm.py:60]: This does not affect anything besides `exp` commands. Internally we use merges when applying stash (merge) commits, which happens both on `exp apply` and `exp run`. When running experiments we generate the stash commit to store the experiment changes, and then apply (merge) them when the experiment is actually executed (either in the workspace or in the separate temp directory).

Comment (jorgeorpinel) [reply dvc/scm.py:60]: OK so I'd make this more specific then. Maybe

```suggestion
                "https://dvc.org/doc/user-guide/troubleshooting#git-shallow"
            )
            msg = (
                f"{msg}:`dvc exp` doesn't work in shallow Git repos. "
                f"See {url} for more information."
```

File: dvc/scm.py [hunk 2]
```diff
@@ -47,6 +49,25 @@ def __init__(self, reason: str) -> None:
         super().__init__(f"{reason}\n{doc}")
 
 
+class GitMergeError(SCMError):
+    def __init__(self, msg: str, scm: Optional["Git"] = None) -> None:
+        if scm and self._is_shallow(scm):
+            url = format_link(
+                "https://dvc.org/doc/user-guide/troubleshooting#git-shallow"
```
Comment (casperdcl) [inline dvc/scm.py:56]: depends on https://github.com/iterative/dvc.org/pull/3577

Comment (skshetry) [reply dvc/scm.py:56]: Also you can use `error_link("git-shallow")` instead of `format_link`.

**Pull request merged**